view/scene.py

```diff
@@ -54,4 +63,77 @@ def plot_quivers(self):
 		
 		plt.axis([-10, 10, -10, 10])
 		self.fig.canvas.draw()
+
+class ContinuousPlotter(object):
+	def __init__(self, get_next_frames_callback):
+		self.get_next_frames_callback = get_next_frames_callback
+
+		self.dt = 0.5
+		self.frame_batch_duration = 10 * self.dt
+		self.current_time = 0.0
+		self.last_update_time = None
+		self.update_period = self.frame_batch_duration
+
+		assert(self.update_period <= self.frame_batch_duration)
+
+		self.max_queue_size = 10
+
+		self.frame_buffer_queue = queue.Queue( maxsize = self.max_queue_size )
+
+		#need this to be atomic boolean (thread safe read/write)
+		self.fetch_more_frames = AtomicVar(True)
+
+	def fetch_continuously(self):
+		frame_count = 0
+		loop_count = 0
+		while True:
+			if not self.fetch_more_frames.getValue():
+				#print ("don't need more frames")
+				self.sleep_after_update()
+				continue
+
+			print("fetching because frame que has size", self.frame_buffer_queue.qsize())
+
+			self.last_update_time = time.time()
+			start_new_frame_callback = time.time()
+			print ("fetching more frames")
+			new_frames = self.get_next_frames_callback(self.dt, self.frame_batch_duration)
+			finish_new_frame_callback = time.time()
+
+			self.frame_buffer_queue.put(new_frames)
+			self.sleep_after_update()
+
+
+	def plot_continuously(self):
+		self.scene = Scene()
+		self.scene.initialize_view()
+		frame_count = 0
+		max_size = 0
+		while True:
+			buffer_queue_size = self.frame_buffer_queue.qsize()
+			print ("buffer queue size", buffer_queue_size)
+			self._check_need_for_frames()
+
+			new_frames = self.frame_buffer_queue.get()
+			for frame in new_frames:
+				new_time, charge_position, field_samples = frame
+
+				#sleep between rendering frames
+				time.sleep((new_time - self.current_time) / fast_forward_rate)
+				self.current_time = new_time
+
+				self.scene.update_view(field_samples, charge_position)
+				self.scene.plot_quivers()
+		
+	def _check_need_for_frames(self):
```

Comment:
hmn... when I read `_check...`, I thought it returns boolean LOL... damn python doesn't support return type. But yeah, probably just return boolean here, then set the variable outside based on return type? So the guys call it "separating query from command". It's a rule in writing function i.e. a function either "get stuff" or "perform action" (an action to mutate the state of the object or the program). Not always applicable, but I tend to stick to the guideline whenever possible. I did ran into situation at work where I have to mix get-command action together though e.g. if you want to verify security access (throw exception on violation), you have to get some security credentials first. In this case, the action of getting security credentials is so tightly coupled with verifying security (get is pre-requisite to command), hence I had no choice but to mix them 😭 (thought very hard on how to seprate them, but in the end, ended up coupling the get and the command action.